### PR TITLE
Align Qwen3-ASR whisper features to the centered-STFT convention (#3535)

### PR DESCRIPTION
**Pull request by eolivelli**
Towards #3535 (Qwen3-ASR 1.7B support): the reason a 1.7B export produced with the same recipe as the official 0.6B collapses into repetition loops on some sung inputs is a feature-frontend mismatch in sherpa-onnx, not the model and not the export.

## Root cause

The whisper-style feature frontend (kaldi-native-fbank `WhisperFeatureComputer`, kaldi `snip_edges=false` placement) centers analysis frame `i` at `i * frame_shift + frame_shift / 2`. The reference extractor these models were trained against (`torch.stft` with `center=True`) centers it at `i * frame_shift`. Every sherpa frame is therefore **half a frame shift (5 ms) late**, and the produced log-mel differs from the reference by **0.064 mean absolute error** (features span roughly [-0.67, 1.33]; Pearson 0.977). Shifting the reference analysis by exactly +80 samples reproduces sherpa's features to 1e-5 — the discrepancy is precisely this offset.

Whisper itself is robust to that. Qwen3-ASR 1.7B is not: on reproducible real-world windows (sung vocals, 12–14.6 s, 16 kHz), greedy decode collapses into a 2–3-token repetition loop that runs to `max_new_tokens`, while the official reference implementation (`qwen-asr` pip package, fp32, transformers backend) transcribes the same float samples cleanly.

Evidence that isolates the cause (all measured on the same input samples):

| pipeline | features | result |
|---|---|---|
| official `qwen-asr` fp32 reference | reference | clean (~25–31 tokens) |
| exported ONNX decoder, teacher-forced vs reference model | reference | **bit-exact logits** (max abs diff 0.0000 over all decode steps) |
| ONNX pipeline outside sherpa (export repo's Python inference) | reference | clean |
| sherpa-onnx, prompt verified token-identical to the reference chat template | sherpa (5 ms late) | **loops** (128 tokens, 2–3 distinct), int8 **and** fp32 decoder |
| sherpa-onnx with waveform pre-shifted 80 samples (alignment emulation) | aligned | clean, token-identical to reference |

The true model has healthy top-1/top-2 logit margins (3–14) on these windows; the shifted spectrogram is a large perturbation, not epsilon noise.

## Fix

1. `WhisperTag.align_to_stft_center` (opt-in, default off): reflect-pad half a frame shift of lead-in before the first chunk and keep exactly `floor(num_samples / frame_shift)` frames. This realigns every frame midpoint to the centered-STFT convention; sherpa's features then match the reference extractor to **2e-5 mean absolute error** (only the first frame differs, from the shorter reflect lead-in). Only the Qwen3-ASR recognizer sets the flag, so Whisper-family models keep their historical features bit-for-bit.

2. A degenerate-repetition guard in the Qwen3-ASR generation loop: if the last 64 generated tokens contain ≤3 distinct ids, decoding has collapsed and cannot recover under greedy sampling; stop early and trim the repetition, returning the text decoded before the collapse. A few inputs sit close enough to the loop attractor that int8 quantization noise alone can tip them over even with exact features (measured: one window loops through the untouched export at int8 with reference features while fp32 is clean); this bounds the damage to an honest truncation instead of a window full of one syllable.

## Validation

- Both reproducible loop windows (extracted from a real song's vocal stem via the downstream pipeline that hit this, exact float samples): previously 128 tokens / 2–3 distinct in every configuration; now clean — one decodes token-identically to the fp32 reference at both int8 and fp32, the other returns the pre-collapse text via the guard instead of garbage.
- A second symptom also fixed: without a language hint one failing window used to return an empty result (the model emitted the language prefix then EOS on the shifted features); it now transcribes.
- 0.6B regression on the official artifact's own 15 `test_wavs` against its bundled `transcript.txt`: mean CER 0.2228 (before) → **0.2109** (after); per-file movement in both directions, no loops, no systematic degradation.
- Whisper recognizers, spoken-language-id, and every other `WhisperTag` consumer are unaffected (flag defaults to off).

The deeper home for the alignment would be kaldi-native-fbank's `WhisperFeatureComputer` itself; this PR keeps the change local, opt-in, and shippable without a dependency bump.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Qwen3-ASR decoding to detect repetitive generation loops, trim redundant output, and stop gracefully.
  * Improved Whisper-compatible transcription timing by aligning audio features with centered short-time analysis windows.
  * Reduced inaccurate trailing audio frames, helping produce cleaner and more reliable transcriptions across streaming input chunks.

### CODE REVIEW

File: sherpa-onnx/csrc/offline-recognizer-qwen3-asr-impl.cc [hunk 1]
```diff
@@ -302,6 +302,79 @@ inline float TensorAbsMax(const Ort::Value &t, int64_t limit) {
...
+        in_loop_set = true;
+        break;
+      }
+    }
+    if (!in_loop_set) {
+      break;
+    }
+    --keep;
+  }
+  ids->resize(keep);
```
[the remark this answers is not in the document]
Comment (eolivelli) [reply sherpa-onnx/csrc/offline-recognizer-qwen3-asr-impl.cc:378]: Good catch — fixed in e4656248. The backward scan is now bounded at the start of the 64-token evidence window, so text preceding the collapse is preserved even when it ends with a token id the loop reuses. Detection fires at the first step the trailing window becomes degenerate, so the loop run barely exceeds the window and the bounded scan still removes it in practice. Re-validated on both reproducible loop windows: behavior unchanged (one clean transcription, one trimmed to empty because its loop starts at token 0).

**Review by csukuangfj — APPROVED** [review]
Thank you for your contribution!
